File: 110.BalancedBinaryTree/BBT.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isBalanced(TreeNode root) {
        return root == null ? true : (getHeight(root) == -1 ? false : true);
    }

    public int getHeight(TreeNode root) {
        if (root == null) {
            return 0;
        }

        if (root.left == null && root.right == null) {
            return 1;
        }

        int lLeft = getHeight(root.left);
        if (lLeft == -1) {
            return -1;
        }

        int lRight = getHeight(root.right);
        if (lRight == -1) {
            return -1;
        }

        if (Math.abs(lLeft - lRight) > 1) {
            return -1;
        }

        return Math.max(lLeft, lRight) + 1;
    }
}

File: 110.BalancedBinaryTree/bbt.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        return root == NULL || getHeight(root) != -1 ? true : false;
    }

    int getHeight(TreeNode* root) {
        if (root == NULL) {
            return 0;
        }

        if (root->left == NULL && root->right == NULL) {
            return 1;
        }

        int l_left = getHeight(root->left);
        if (l_left == -1) {
            return -1;
        }

        int l_right = getHeight(root->right);
        if (l_right == -1) {
            return -1;
        }

        if (fabs(l_left - l_right) > 1) {
            return -1;
        }

        return max(l_left, l_right) + 1;
    }
};
